app/components/IconPicker/components/EmojiPanel.tsx

```diff
@@ -0,0 +1,251 @@
+  sortFrequencies,
+} from "../utils";
+import GridTemplate, { DataNode } from "./GridTemplate";
+import SkinPicker from "./SkinPicker";
+
+/**
+ * This is needed as a constant for react-window.
+ * Calculated from the heights of TabPanel and InputSearch.
+ */
+const GRID_HEIGHT = 362;
+
+const useEmojiState = () => {
+  const [emojiSkin, setEmojiSkin] = usePersistedState<EmojiSkin>(
+    getEmojiSkinKey(),
+    EmojiSkin.Default
+  );
+  const [emojisFreq, setEmojisFreq] = usePersistedState<Record<string, number>>(
+    getEmojisFreqKey(),
+    {}
+  );
+  const [lastEmoji, setLastEmoji] = usePersistedState<string | undefined>(
+    getLastEmojiKey(),
+    undefined
+  );
+
+  const incrementEmojiCount = React.useCallback(
+    (emoji: string) => {
+      emojisFreq[emoji] = (emojisFreq[emoji] ?? 0) + 1;
+      setEmojisFreq({ ...emojisFreq });
+      setLastEmoji(emoji);
+    },
+    [emojisFreq, setEmojisFreq, setLastEmoji]
+  );
+
+  const getFreqEmojis = React.useCallback(() => {
+    const freqs = Object.entries(emojisFreq);
+
+    if (freqs.length > FREQUENTLY_USED_COUNT.Track) {
+      sortFrequencies(freqs).splice(FREQUENTLY_USED_COUNT.Track);
+      setEmojisFreq(Object.fromEntries(freqs));
+    }
+
+    const emojis = sortFrequencies(freqs)
+      .slice(0, FREQUENTLY_USED_COUNT.Get)
+      .map(([emoji, _]) => emoji);
+
+    const isLastPresent = emojis.includes(lastEmoji ?? "");
+    if (lastEmoji && !isLastPresent) {
+      emojis.pop();
+      emojis.push(lastEmoji);
+    }
+
+    return emojis;
+  }, [emojisFreq, setEmojisFreq, lastEmoji]);
+
+  return {
+    emojiSkin,
+    setEmojiSkin,
+    incrementEmojiCount,
+    getFreqEmojis,
+  };
+};
+
+type Props = {
+  panelWidth: number;
+  query: string;
+  panelActive: boolean;
+  onEmojiChange: (emoji: string) => void;
+  onQueryChange: (query: string) => void;
+};
+
+const EmojiPanel = ({
+  panelWidth,
+  query,
+  panelActive,
+  onEmojiChange,
+  onQueryChange,
+}: Props) => {
+  const { t } = useTranslation();
+
+  const searchRef = React.useRef<HTMLInputElement | null>(null);
+  const scrollableRef = React.useRef<HTMLDivElement | null>(null);
+
+  const {
+    emojiSkin: skin,
+    setEmojiSkin,
+    incrementEmojiCount,
+    getFreqEmojis,
+  } = useEmojiState();
+
+  const freqEmojis = React.useMemo(() => getFreqEmojis(), [getFreqEmojis]);
+
+  const handleFilter = React.useCallback(
+    (event: React.ChangeEvent<HTMLInputElement>) => {
+      onQueryChange(event.target.value);
+    },
+    [onQueryChange]
+  );
+
+  const handleSkinChange = React.useCallback(
+    (emojiSkin: EmojiSkin) => {
+      setEmojiSkin(emojiSkin);
+    },
+    [setEmojiSkin]
+  );
+
+  const handleEmojiClick = React.useCallback(
+    ({ id, value }: { id: string; value: string }) => {
+      onEmojiChange(value);
+      incrementEmojiCount(id);
+    },
+    [onEmojiChange, incrementEmojiCount]
+  );
+
+  const isSearch = query !== "";
+  const templateData: DataNode[] = isSearch
```

Comment:
`getAllEmojis` can be memoized. I don't think search can be.. 🤔

Reply:
That'd just be unnecessary, I think.
The search flow gets executed only when query is non-empty, and the component has to re-render anyway whenever the query changes.
Am I missing something? 🤔

Reply:
In the case of changing query would bust yea so that part isn't needed to be within memo, just thought it might make dx easier to wrap the whole of `templateData`. Whatever you choose 👍

Reply:
On second thought, we do not `memo` for either of the cases since they are mutually exclusive (and needs component re-render).